dex/encrypt/encrypt.go

```diff
@@ -0,0 +1,232 @@
+}
+
+// Deserialize deserializes the encryption key for the password.
+func Deserialize(pw string, encCrypter []byte) (Crypter, error) {
+	ver, pushes, err := encode.DecodeBlob(encCrypter)
+	if err != nil {
+		return nil, err
+	}
+	switch ver {
+	case 0:
+		return decodeArgonPoly_v0(pw, pushes)
+	default:
+		return nil, fmt.Errorf("unknown Crypter version %d", ver)
+	}
+}
+
+// argonPolyCryper is an encryption algorithm based on argon2id for key
+// derivation and xchacha20poly1305 for symmetric encryption.
+type argonPolyCrypter struct {
+	key    Key
+	salt   Salt
+	params *argonParams
+}
+
+// newArgonPolyCrypter is the constructor for an argonPolyCrypter.
+func newArgonPolyCrypter(pw string) *argonPolyCrypter {
+	pwB := []byte(pw)
+	salt := newSalt()
+	threads := uint8(runtime.NumCPU())
+	keyB := argon2.IDKey(pwB, salt[:], defaultTime, defaultMem, threads, KeySize)
+	var key Key
+	copy(key[:], keyB)
+	return &argonPolyCrypter{
+		key:  key,
+		salt: salt,
+		params: &argonParams{
+			time:    defaultTime,
+			memory:  defaultMem,
+			threads: threads,
+		},
+	}
+}
+
+// Encrypt encrypts the plaintext.
+func (c *argonPolyCrypter) Encrypt(plainText []byte) ([]byte, error) {
+	keyHash := blake256.Sum256(c.key[:])
+	boxer, err := chacha20poly1305.NewX(c.key[:])
+	if err != nil {
+		return nil, fmt.Errorf("aead error: %v", err)
+	}
+	nonceSize := boxer.NonceSize()
+	nonce := make([]byte, nonceSize)
+	_, err = rand.Read(nonce)
+	if err != nil {
+		return nil, fmt.Errorf("nonce generation error: %v", err)
+	}
+	cipherText := boxer.Seal(nil, nonce, plainText, keyHash[:])
+	return encode.BuildyBytes{0}.AddData(nonce).AddData(cipherText), nil
+}
+
+// Decrypt decrypts the ciphertext created by Encrypt.
+func (c *argonPolyCrypter) Decrypt(encrypted []byte) ([]byte, error) {
+	ver, pushes, err := encode.DecodeBlob(encrypted)
+	if err != nil {
+		return nil, fmt.Errorf("DecodeBlob: %v", err)
+	}
+	if ver != 0 {
+		return nil, fmt.Errorf("only version 0 encryptions are known. got version %d", ver)
+	}
+	if len(pushes) != 2 {
+		return nil, fmt.Errorf("expected 2 pushes. got %d", len(pushes))
+	}
+	boxer, err := chacha20poly1305.NewX(c.key[:])
+	if err != nil {
+		return nil, fmt.Errorf("aead error: %v", err)
+	}
+	nonce, cipherText := pushes[0], pushes[1]
+	if len(nonce) != boxer.NonceSize() {
+		return nil, fmt.Errorf("incompatible nonce length. expected %d, got %d", boxer.NonceSize(), len(nonce))
+	}
+	keyHash := blake256.Sum256(c.key[:])
+	plainText, err := boxer.Open(nil, nonce, cipherText, keyHash[:])
+	if err != nil {
+		return nil, fmt.Errorf("aead.Open: %v", err)
+	}
+	return plainText, nil
+}
+
+// Serialize serializes the argonPolyCrypter. It is an error to call Serialize
+// on an un-initialized or a closed argonPolyCrypter.
+func (c *argonPolyCrypter) Serialize() ([]byte, error) {
+	if c.key == zeroKey {
+		return nil, fmt.Errorf("cannot Serialize a closed argonPolyCrypter")
+	}
+	ckSum, err := c.Encrypt([]byte(checkPhrase))
```

Comment:
i would not encrypt a known plaintext, because this unnecessarily reveals these bits from the keystream.

what's the point in doing this?

Reply:
Yeah. Wasn't sure if this would fly or not. 

When re-creating the key from the password and the KDF params & salt, do you check that the password is the same password used to create the key associated with serialized parameters? I guess you'd find out later when you went to decrypt something, but it'd be nice to know earlier. 

Removing this check.

Reply:
There are a couple of approaches to this.  In snacl, it records the sha256 of the derived key and compares to that, but I dislike this way because it will make any usage of a hash of that secret key insecure if used elsewhere.

In ss for the passphrase encryption streams, I end up deriving a 64 byte key from Argon2id.  I use the first 32 bytes as a poly1305 authentication key, and the last 32 bytes as the symmetric encryption key.

https://github.com/jrick/ss/blob/a6864ccdc030e0976d62b3efef4bdbb48104282c/stream/stream.go#L121-L137